Add MATLAB PID design toolchain for motor control

CesMag opened the pull request:
## Summary
- Add MATLAB scripts for PID controller design based on DFRobot FIT0520 motor specs
- Motor transfer function model: G(s) = 3.665 / (0.05s + 1)
- Computed velocity PID gains: Kp=0.159, Ki=7.03, Kd=0
- Includes cascaded PID design for current→velocity→position control
- System identification template ready for real encoder data

## Motor Parameters (Verified)
| Parameter | Value |
|-----------|-------|
| Rated Voltage | 6V |
| No-load Speed | 210 RPM @ 0.13A |
| Stall Current | 3.2A |
| Stall Torque | 10 kg·cm |
| Encoder PPR | 341.2 (11 × 34.02) |

## Scripts Added
- `motor_params.m` - Motor specifications
- `motor_model_1st_order.m` - First-order TF model
- `motor_model_2nd_order.m` - Full second-order model
- `pid_design_velocity.m` - Auto-tuning with pidtune()
- `pid_discretize.m` - 100 Hz discretization for ESP32
- `simulate_step_response.m` - Closed-loop simulation
- `cascaded_pid_design.m` - Cascade control design
- `system_id_template.m` - System ID from encoder data

## Test Plan
- [x] All scripts verified working with MATLAB R2025a
- [ ] Integrate computed gains into ESP32 star_pid library
- [ ] Collect real encoder step response data
- [ ] Run system_id_template.m to refine time constant

CesMag committed:
Add MATLAB PID design toolchain for motor control

Motor: DFRobot FIT0520 (6V, 210RPM, 3.2A stall, 341.2 PPR)

Scripts:
- motor_params.m: Motor specifications and derived parameters
- motor_model_1st_order.m: G(s) = 3.665 / (0.05s + 1)
- motor_model_2nd_order.m: Full model with state-space representation
- pid_design_velocity.m: Auto-tuning @ 20 rad/s bandwidth
- pid_discretize.m: 100 Hz discretization for ESP32
- simulate_step_response.m: Closed-loop simulation
- cascaded_pid_design.m: Current→Velocity→Position cascade
- system_id_template.m: Template for encoder data fitting

Computed velocity PID: Kp=0.159, Ki=7.03, Kd=0
Performance: Rise=70ms, Settle=240ms, Overshoot=5.8%

CesMag committed:
Address PR review comments

- motor_params.m: Fix I_no_load comment (from datasheet), add inductance
  measurement guidance, calculate viscous friction from no-load data,
  add wheel inertia calculation with mass estimate
- simulate_step_response.m: Fix disturbance model to use sensitivity
  function S=1/(1+GC), clarify disturbance units as rad/s
- pid_config_output.txt: Fix integral limits to ±50 (50% of output range)
- system_id_template.m: Replace deprecated datestr(now) with datetime()
- identified_params.txt: Add note that this is example/demo output
- cascaded_pid_design.m: Clarify that velocity loop output limits require
  current sensing, reference velocity-only config as alternative
- motor_model_1st_order.m: Add note about 50ms time constant estimate source

bsikar commented:
Thanks for the updates! I can see you addressed several documentation items (cascaded PID clarifications, example data notes, etc.) which look good. However, the critical blocking issue with the motor time constant has not been fixed.

  The Problem

  The PR still uses tau_est = 0.05 (50ms), but our own design document explicitly uses 75ms for all calculations:

  Evidence from docs/Protobuf_Protocol_Design_Analysis.tex:

  1. Line 626 (control frequency calculation):
  f_{control} ≥ 10/τ_m = 10/0.075s ≈ 133 Hz minimum
  1. Uses 0.075s = 75ms ✅
  2. Line 1022 (evolution table):
  Motor time constant: 50~ms (assumed) → ~75~ms (estimated)
  2. Explicitly states 75ms is the ESTIMATED value ✅
  3. Line 729 (decision rationale):
  Motor dynamics: 4.7× time constant (good margin)
  3. At 100 Hz: 10ms / 75ms = 7.5× ≈ "4.7×" stated ✅

  Why This Matters

  The 50% difference in time constant means:
  - Transfer function: G(s) = 3.665/(0.05s+1) should be G(s) = 3.665/(0.075s+1)
  - Natural bandwidth: 20 rad/s should be 13.3 rad/s
  - All PID gains (Kp=0.159, Ki=7.03) are computed for the wrong plant

  When we test on real hardware with these gains, we'll likely see oscillation/overshoot because we designed PID for a motor that's 50% faster than reality.

  Required Fix

  In matlab/motor_model_1st_order.m, line 27, change:

  % BEFORE:
  % Initial estimate of 50ms based on typical geared DC motor values (literature: 30-100ms)
  tau_est = 0.05;  % Estimated 50ms - MEASURE THIS ON REAL MOTOR

  % AFTER:
  % Time constant from Protobuf_Protocol_Design_Analysis.tex (line 622, 1022)
  % Design doc uses 75ms for all calculations (control frequency, margins)
  % This matches "estimated" value, not "assumed" value from evolution table
  tau_est = 0.075;  % Estimated 75ms from design doc - MEASURE THIS ON REAL MOTOR

  Then re-run:
  1. motor_model_1st_order.m
  2. motor_model_2nd_order.m
  3. pid_design_velocity.m
  4. cascaded_pid_design.m
  5. simulate_step_response.m

  And update:
  - README.md with new computed gains
  - PR description with corrected transfer function G(s) = 3.665/(0.075s+1)

  Why Not Use 50ms?

  The added comment cites "literature: 30-100ms" but this is irrelevant because:
  - We already analyzed this and chose 75ms in our design document
  - Our communication protocol design (SPI timing, bandwidth margins) all use 75ms
  - We can't have one value in the design doc and a different one in code
  - When system ID runs on real hardware, we'll refine from 75ms → actual (not from 50ms)

  Using 75ms is the conservative choice - better to design for a slower motor than a faster one.

  Timeline

  This is a 30-minute fix (re-run scripts + update docs) that will save hours of debugging when we test on hardware.

  ---
  Everything else in this PR is excellent work. The structure, documentation, and workflow are all top-notch. This is just a one-line parameter correction to match our existing design document.

  Once this is fixed, immediate approval. Let me know if you need any clarification!

CesMag committed:
Fix motor time constant to 75ms per design doc

Critical fix: Changed tau from 50ms to 75ms to match
docs/Protobuf_Protocol_Design_Analysis.tex (lines 622, 1022)

Updated transfer function: G(s) = 3.665 / (0.075s + 1)

New computed PID gains:
- Velocity-only: Kp=0.2856, Ki=8.0084, Kd=0
- Performance: Rise=70ms, Settle=260ms, Overshoot=9.7%

Also fixed integral_limit calculation bug in pid_discretize.m
(was computing 100 instead of 50 for anti-windup)
